src/plugins/discover/public/application/apps/main/services/use_saved_search.ts

```diff
@@ -0,0 +1,337 @@
+import { SortOrder } from '../../../../saved_searches/types';
+import { getDimensions, getChartAggConfigs } from '../utils';
+import { buildPointSeriesData, Chart } from '../components/chart/point_series';
+import { TimechartBucketInterval } from '../components/timechart_header/timechart_header';
+import { useSingleton } from '../utils/use_singleton';
+import { FetchStatus } from '../../../types';
+
+export type SavedSearchDataSubject = BehaviorSubject<SavedSearchDataMessage>;
+export type SavedSearchRefetchSubject = Subject<SavedSearchRefetchMsg>;
+
+export interface UseSavedSearch {
+  refetch$: SavedSearchRefetchSubject;
+  data$: SavedSearchDataSubject;
+}
+
+export type SavedSearchRefetchMsg = 'reset' | undefined;
+
+export interface SavedSearchDataMessage {
+  bucketInterval?: TimechartBucketInterval;
+  chartData?: Chart;
+  fetchCounter?: number;
+  fetchError?: Error;
+  fieldCounts?: Record<string, number>;
+  hits?: number;
+  inspectorAdapters?: { requests: RequestAdapter };
+  rows?: ElasticSearchHit[];
+  state: FetchStatus;
+}
+
+/**
+ * This hook return 2 observables, refetch$ allows to trigger data fetching, data$ to subscribe
+ * to the data fetching
+ * @param indexPattern
+ * @param savedSearch
+ * @param searchSessionManager
+ * @param searchSource
+ * @param services
+ * @param state
+ * @param stateContainer
+ * @param useNewFieldsApi
+ */
+export const useSavedSearch = ({
+  indexPattern,
+  savedSearch,
+  searchSessionManager,
+  searchSource,
+  services,
+  state,
+  stateContainer,
+  useNewFieldsApi,
+}: {
+  indexPattern: IndexPattern;
+  savedSearch: SavedSearch;
+  searchSessionManager: DiscoverSearchSessionManager;
+  searchSource: SearchSource;
+  services: DiscoverServices;
+  state: AppState;
+  stateContainer: GetStateReturn;
+  useNewFieldsApi: boolean;
+}): UseSavedSearch => {
+  const { data, filterManager, uiSettings } = services;
+  const timefilter = data.query.timefilter.timefilter;
+
+  const initFetchState: FetchStatus = useMemo(() => {
+    // A saved search is created on every page load, so we check the ID to see if we're loading a
+    // previously saved search or if it is just transient
+    const shouldSearchOnPageLoad =
+      uiSettings.get<boolean>(SEARCH_ON_PAGE_LOAD_SETTING) ||
+      savedSearch.id !== undefined ||
+      timefilter.getRefreshInterval().pause === false ||
+      searchSessionManager.hasSearchSessionIdInURL();
+    return shouldSearchOnPageLoad ? FetchStatus.LOADING : FetchStatus.UNINITIALIZED;
+  }, [uiSettings, savedSearch.id, searchSessionManager, timefilter]);
+
+  /**
+   * The observable the UI (aka React component) subscribes to get notified about
+   * the changes in the data fetching process (high level: fetching started, data was received)
+   */
+  const data$: SavedSearchDataSubject = useSingleton(
+    () =>
+      new BehaviorSubject({
+        state: initFetchState as FetchStatus,
+      })
```

Comment:
:art: That way we'll avoid the cast (which is from a typing perspective, way more aggresive and error-prone than specifying the generic).

```suggestion
      new BehaviorSubject<SavedSearchDataMessage>({
        state: initFetchState,
      })
```

I also find this curious that it's happening at all... We define `initFetchState` as `FetchStatus` above, but somehow TS feels the need to outsmart us here again, by saying: "well, this is actually `FetchStatus.LOADING | FetchStatus.UNINITIALIZED` even though you already told me to use the more generic type!"